prog.go

```diff
@@ -549,6 +549,16 @@ func (p *Program) Close() error {
 	return p.fd.Close()
 }
 
+// Extented Program test options and attributes
+type TestOptions struct {
+	In     []byte
+	Ctx    []byte
+	CtxOut []byte
```

Comment:
I'm not an expert on go, but how would you convert from `[]byte` -> `interface{}` for returned context?

Reply:
When working with converting Go types to/from binary, my mind immediately goes to `encoding.Binary(Un)Marshaler`. We could take a `BinaryMarshaler` for `Context`.

The following needs to be taken into account:
```
 *		**BPF_PROG_TYPE_SK_LOOKUP**
 *			*data_in* and *data_out* must be NULL.
 *
 *		**BPF_PROG_TYPE_RAW_TRACEPOINT**,
 *		**BPF_PROG_TYPE_RAW_TRACEPOINT_WRITABLE**
 *
 *			*ctx_out*, *data_in* and *data_out* must be NULL.
 *			*repeat* must be zero.
```

With the current patch, I don't think these prog types will work, since all fields (data_in/out, ctx_in/out) unconditionally receive slice pointers, even if their lengths are zero. This will need to be changed to conditionally set the respective fields when their input slices are > len 0.

On the bright side, this constraint might guide us in designing the API:

```go
type RunContext struct {
	// Program's data input. Required field.
	DataIn []byte

	// Program's context input. Optional field.
	ContextIn encoding.BinaryMarshaler

	// Program's context after Program has run. Optional field.
	ContextOut encoding.BinaryUnmarshaler
	// .. or
	ContextOut []byte
// ...
}
```

- DataOut could be the return value of the function or caller-provided. Both could work given the input struct is called `RunContext`. The forced allocation of the output buffer w/ padding could be moved to `Test()` and we don't need to break API. `Benchmark()` could then run with SkipOutput implicitly enabled.
- If ContextIn is provided, a buffer for ContextOut can be allocated with the same length as the marshaling result. If ContextOut is not provided, this buffer would not be allocated and the CtxOut arg not provided to the syscall to accomodate for `RAW_TRACEPOINT*`.

Reply:
With your proposed `RunContext`, we still need to give a way for the user to provide the context value, right? `ContextIn` and `ContextOut` are interfaces. We still need the data somewhere. For that reason, I think it might be cleaner to have `Context` and `ContextOut` to be `interface{}`.

> With the current patch, I don't think these prog types will work, since all fields (data_in/out, ctx_in/out) unconditionally receive slice pointers, even if their lengths are zero.

I think even without this patch, BPF_PROG_TYPE_RAW_TRACEPOINT* was broken: https://github.com/cilium/ebpf/blob/e0825791eaf28e70c8c6403054c85b5d37450290/prog.go#L585-L587 . `data_in` was never permitted to be null.

Reply:
I'd like to also better understand the concern about allocating for `Benchmark()`. The prog iterations are run inside the kernel, so allocation time is not measured.

Reply:
> When working with converting Go types to/from binary, my mind immediately goes to encoding.Binary(Un)Marshaler.

I'm not a fan of either interface. Because they are usually implemented manually they invite bugs and bloat the code base. BinaryMarshaler also doesn't allow controlling allocation, which really should be possible if the reason to use BinaryMarshaler is performance in the first place.

Hence, I think that interface{} is actually better than the marshalers, since it allows us to cut out all that junk in the common case.

> I think even without this patch, BPF_PROG_TYPE_RAW_TRACEPOINT* was broken.

That's correct. I think Timo is saying that our new API should allow passing NULL in some fashion. We can do that as a follow up PR though, I think.

> I'd like to also better understand the concern about allocating for Benchmark().

The issue is that we allocate an output slice even if the caller discards it: https://github.com/cilium/ebpf/blob/e0825791eaf28e70c8c6403054c85b5d37450290/prog.go#L602